packages/core/src/runtime/AbstractRuntime.ts

```diff
@@ -222,49 +206,146 @@ export abstract class AbstractRuntime {
 		}
 
-		const keyHash = AbstractRuntime.getKeyHash(key)
-		const lowerBound = `${model}/${keyHash}/${MIN_MESSAGE_ID}`
-		let upperBound = `${model}/${keyHash}/${MAX_MESSAGE_ID}`
+		const visited = new Set()
 
 		// eslint-disable-next-line no-constant-condition
 		while (true) {
-			const results = await this.db.query<{ key: string; value: Uint8Array; clock: number }>("$effects", {
-				where: { key: { gte: lowerBound, lt: upperBound } },
-				orderBy: { key: "desc" },
+			const currentMessagePosition = stack.pop()
+			// no more messages to visit
+			if (!currentMessagePosition) {
+				break
+			}
+			visited.add(`${currentMessagePosition.branch}/${currentMessagePosition.clock}`)
+
+			const effectsOnThisBranch = await this.db.query<{
+				key: string
+				value: Uint8Array | undefined
+				branch: number
+				clock: number
+			}>("$effects", {
+				where: {
+					key: { gte: lowerBound, lt: upperBound },
+					branch: currentMessagePosition.branch,
+					clock: { lte: currentMessagePosition.clock },
+				},
+				orderBy: { clock: "desc" },
 				limit: 1,
 			})
 
-			if (results.length === 0) {
-				return null
+			const parentPositions: Position[] = []
+			const matchingEffectOnThisBranch = effectsOnThisBranch[0]
+			if (matchingEffectOnThisBranch) {
+				if (matchingEffectOnThisBranch.clock == currentMessagePosition.clock) {
+					// the current message is the latest effect on this branch
+					result.add(matchingEffectOnThisBranch.key)
+					// don't explore this message's parents
+					continue
+				} else {
+					parentPositions.push({ branch: matchingEffectOnThisBranch.branch, clock: matchingEffectOnThisBranch.clock })
+				}
 			}
 
-			const [{ key: effectKey, value }] = results
-			const [{}, {}, messageId] = effectKey.split("/")
+			// check for branches that merge into this branch
+			const branchMergeQuery = {
+				where: {
+					target_branch: currentMessagePosition.branch,
+					target_clock: { lte: currentMessagePosition.clock },
+				},
+			}
+			if (matchingEffectOnThisBranch) {
+				// @ts-ignore
+				branchMergeQuery.where.target_clock.gt = matchingEffectOnThisBranch.clock
+			}
+			for (const branchMerge of await context.messageLog.db.query<BranchMergeRecord>(
+				"$branch_merges",
+				branchMergeQuery,
+			)) {
+				parentPositions.push({
+					branch: branchMerge.source_branch,
+					clock: branchMerge.source_clock,
+				})
+			}
 
-			upperBound = effectKey as string
-			const visited = new Set<string>()
+			for (const parentPosition of parentPositions) {
+				if (!visited.has(`${parentPosition.branch}/${parentPosition.clock}`)) {
+					stack.push(parentPosition)
+				}
+			}
+		}
 
-			for (const parent of context.message.parents) {
-				const isAncestor = await context.messageLog.isAncestor(parent, messageId, visited)
-				if (isAncestor) {
-					return cbor.decode<null | T>(value)
+		// remove messages that are parents of each other
+		const messagesToRemove = new Set<string>()
```

Comment:
I'm unsure if we need to do this at all - if we return some messages that are ancestors of each other, then the result will be the same because the merge function is commutative. The question here is whether it's cheaper to check if messages are parents of each other or cheaper to just call merge a few more times. I think it would depend on the size of the value and the topology of the message graph.